File: benchmarks/polysum/polysum.zig
// Polynomial ranged-sum benchmark — multi-term polynomials (Zig, fixed u64).
//
// IMPORTANT: overflows 2^64 past degree 2 (x^7/x^20 at once) — values are
// mod 2^64 via wrapping ops, so they are WRONG. Native-loop SPEED reference
// only. N/REPS from argv (defaults 1_000_000 / 100).
//
// Targets Zig 0.16: main takes std.process.Init.Minimal (args via
// init.args.iterate()), output via std.posix.write (the std I/O writer
// path needs an explicit Io instance in 0.16). The args/I/O APIs churn
// between Zig versions — this file is 0.16-specific.
//
// Build: zig build-exe -OReleaseFast polysum.zig -femit-bin=polysum

const std = @import("std");

fn ipow(base: u64, e: u32) u64 {
    var r: u64 = 1;
    var i: u32 = 0;
    while (i < e) : (i += 1) r *%= base;
    return r;
}

pub fn main(init: std.process.Init.Minimal) !void {
    var it = init.args.iterate();
    _ = it.next(); // argv[0]
    var n: u64 = 1_000_000;
    var reps: u64 = 100;
    if (it.next()) |a| n = try std.fmt.parseInt(u64, a, 10);
    if (it.next()) |a| reps = try std.fmt.parseInt(u64, a, 10);

    var t1: u64 = 0;
    var t2: u64 = 0;
    var t3: u64 = 0;
    var t7: u64 = 0;
    var t20: u64 = 0;
    var r: u64 = 0;
    while (r < reps) : (r += 1) {
        const lo: u64 = 1 + r;
        const hi: u64 = n + r;
        var x: u64 = lo;
        while (x <= hi) : (x += 1) {
            t1 +%= 2 *% x +% 3;
            t2 +%= 5 *% ipow(x, 2) -% 3 *% x +% 1;
            t3 +%= 4 *% ipow(x, 3) -% 2 *% ipow(x, 2) +% 7 *% x -% 5;
            t7 +%= 92 *% ipow(x, 7) +% 13 *% ipow(x, 3) -% 5 *% x +% 8;
            t20 +%= ipow(x, 20) +% 17 *% ipow(x, 13) -% 4 *% ipow(x, 5) +% 2 *% x +% 9;
        }
    }

    var buf: [256]u8 = undefined;
    const s = try std.fmt.bufPrint(&buf, "{d}\n{d}\n{d}\n{d}\n{d}\n", .{ t1, t2, t3, t7, t20 });
    _ = std.c.write(std.posix.STDOUT_FILENO, s.ptr, s.len);
}

File: benchmarks/polysum/polysum.rs
// Polynomial ranged-sum benchmark — multi-term polynomials (Rust, fixed u64).
//
// IMPORTANT: overflows 2^64 past degree 2 (x^7/x^20 at once). Uses
// std::num::Wrapping so it computes mod 2^64 in both debug and release —
// the printed values are WRONG. Native-loop SPEED reference only.
// N/REPS from argv (defaults 1_000_000 / 100).
//
// Build: rustc -O polysum.rs -o polysum && ./polysum

use std::env;
use std::num::Wrapping;

type W = Wrapping<u64>;

fn ipow(base: W, e: u32) -> W {
    let mut r = Wrapping(1u64);
    for _ in 0..e {
        r *= base;
    }
    r
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let n: u64 = if args.len() > 1 { args[1].parse().unwrap() } else { 1_000_000 };
    let reps: u64 = if args.len() > 2 { args[2].parse().unwrap() } else { 100 };

    let (mut t1, mut t2, mut t3, mut t7, mut t20) =
        (Wrapping(0u64), Wrapping(0u64), Wrapping(0u64), Wrapping(0u64), Wrapping(0u64));
    let c = |k: u64| Wrapping(k);

    for r in 0..reps {
        let (lo, hi) = (1 + r, n + r);
        let mut xi = lo;
        while xi <= hi {
            let x = Wrapping(xi);
            t1 += c(2) * x + c(3);
            t2 += c(5) * ipow(x, 2) - c(3) * x + c(1);
            t3 += c(4) * ipow(x, 3) - c(2) * ipow(x, 2) + c(7) * x - c(5);
            t7 += c(92) * ipow(x, 7) + c(13) * ipow(x, 3) - c(5) * x + c(8);
            t20 += ipow(x, 20) + c(17) * ipow(x, 13) - c(4) * ipow(x, 5) + c(2) * x + c(9);
            xi += 1;
        }
    }

    println!("{}\n{}\n{}\n{}\n{}", t1.0, t2.0, t3.0, t7.0, t20.0);
}
